### ISSUE TITLE
Veyon for multi-seat systems

### ISSUE THREAD

**Issue by stephaniebamerico** [veyon/veyon#299]
I work for a project of the Brazilian federal government and we are interested in using its tool in our system. I really liked the interface and operation of the system, and I believe it will be of great benefit to us. However, I would need to modify it to allow multiple users on the same machine (multi-terminal system). Do you have any idea how to do this on your system? I wonder if you have something in mind before I start studying the whole solution. Basically, besides the IP of the machine, we would need to insert the exported Display information.

I await your return. Thank you!

**Comment by lbssousa** [comment]
IIRC `veyon-service` already supports an option `-session <NUMBER>` which can be used to shift the default port it will listen to. With this option, one can be able to launch several `veyon-service` processes in a single host, which is the case for multi-seat/multi-session setups.

One would be tempted to use `${XDG_SESSION_ID}` variable as a seed for shifting the port number, but I guess it may lead to an inconsistent numbering of ports, as session numbers change after each login/logout. Instead, one can sort the seat list from `loginctl list-seats` command and take its enumeration as a seed for `-session` option. A similar approach was already done for `italc` package maintained by openSUSE Education:

https://build.opensuse.org/package/rdiff/Education/italc?linkrev=base&rev=30

I guess the missing piece here is a proper way to autostart `veyon-service` correctly for each seat. The current systemd service unit which will be shipped with Veyon 4.1 is not prepared for this case.

**Comment by lbssousa** [comment]
I'm testing Veyon 4.0.5 in an Ubuntu 16.04 multi-seat setup here (authentication keys already installed):
1. I've configured second seat to launch `veyon-service` with option `-session 1`
2. I've added manually the "computers" in `veyon-configurator` with addresses `192.168.x.x:11100` for the first seat, and `192.168.x.x:11101` for the second one.
3. After starting `veyon-master`, I'm able to see correctly the thumbnails of my client's both seats.
4. Per-seat screen locking works correctly.
5. Per-seat master screen broadcasting seems to work correctly.
6. **Per-seat remote control is NOT working correcly!** Regardless I select the first or the second seat to control remotely, it's always the first seat which is captured.

**Comment by lbssousa** [comment]
One additional note: in my tests, in order to get Veyon clients appearing properly in my master, they must be run as *normal user* (e.g. via `/etc/xdg/autostart` file), although Veyon documentation suggests that they should be run as root.

**Comment by tobydox** [comment]
Your findings and remarks are all correct, thank you! There's currently no dedicated multi-seat/-session support in Veyon but we plan to add it in Veyon 4.2, maybe already in 4.1. Until then you'll have to deal with starting the service manually and add use increased port numbers.

Regarding remote control there's probably a bug we need to fix. If a computer is visible as thumbnail it should be accessible via remote control as well. I'll have a look at it.

**Comment by stephaniebamerico** [comment]
The best way to start Veyon automatically on a multiterminal system seems to be to add a _/etc/X11/Xsession.d/10veyon-service_ file with the following content:

```
SEAT=$(loginctl seat-status | head -1)
N=$(($(loginctl list-seats | grep -n $SEAT | cut -d ':' -f1)-2))
veyon-service -session $N disown &
```

This will allow each machine to find its session number correctly.

However, the remote control or even the screen display are not working. I imagine it has already been fixed as there are commits related to the problem.

**Comment by tobydox** [comment]
Thanks for the feedback and the useful information for the session handling! As it doesn't seem to be that complicated chances are good we'll even integrate it in upcoming Veyon 4.1. The remote control issue will be fixed in version 4.0.6 next week :-)

**Comment by lbssousa** [comment]
@stephaniebamerico what's that `disown` argument in your command line example? I have not found it in veyon documentation&sources.

**Comment by lbssousa** [comment]
@stephaniebamerico I've tested your approach in my Xubuntu installation, and noticed that I'm unable to logoff/reboot/shutdown my students' machine from veyon-master when I start veyon-service this way (I got a DBus error regarding `org.xfce.SessionManager`).

The only way I've found to make my veyon remote clients to logoff/reboot/shutdown correctly is spawning `veyon-service` via XDG Autostart Desktop files. I've created a script `/usr/local/bin/veyon-service-launcher` with a slightly modified content from your example:

```bash
#!/bin/bash
exec veyon-service -session $(( $(loginctl list-seats | grep "seat[^s]" | sort | grep -n $XDG_SEAT | cut -d ':' -f1) - 1 ))
```

and a minimal `veyon-service-launcher.desktop` file in `/etc/xdg/autostart` with the following content:

```
[Desktop Entry]
Type=Application
Name=Veyon Service Launcher
TryExec=veyon-service
Exec=veyon-service-launcher
```

**Comment by tobydox** [comment]
Powering off/rebooting should be fixed in upcoming 4.0.91 release as indicated in #312. It's possible you'll still see dbus messages as Veyon simply tries all kind of DBus calls in order to reach at least one session manager. You can ignore KDE messages if you're running Gnome etc.

Besides that I implemented basic multi session support today so with the latest version you can enable "multi session support" in Veyon Configurator which will make the service start server instances for every graphical session with different session IDs (based on an internal list of used/unused session IDs) :-)
